### PR TITLE
Adding new tube status script

### PR DESCRIPTION
**Pull request by ivernus**
Adding a script to eh contrib dir. I made this for my own i3 to keep an eye on the TFL tube status. I tried to get it working with the wrapper script but couldn't manage it.

### CODE REVIEW

**Comment by stapelberg** [comment]
AFAICT, this script sends one request to api.tfl.gov.uk every `interval` (defaulting to 5s but often changed to 1s). I think that’s a bit too much. Could you please change the script to be less aggressive?

**Comment by brioscaibriste** [comment]
Thanks Michael, good point, I will add a time delay option and hopefully re request merge soon.

**Comment by Airblader** [comment]
You don't need to close the pull request, just force push onto the branch and Github will update it. We can wait. :9

**Comment by brioscaibriste** [comment]
Thanks for the tip Airblader :)

**Comment by brioscaibriste** [comment]
Hi folks, just dropped a quick update of the scripts with some throttling in place. It can be tuned to run every X minutes, otherwise caching the line status data in /tmp. I hope this is correct and makes a bit more sense than the original one.

File: contrib/tubestatusdef.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+ two functions at least, one for time, one for polling, returning status value
+'''
+
+def Throttle(PollInterval,Throttle,TFileName):
+
+    # Does the temporary file exist
+    TFileStatus = os.path.isfile(TFileName)
+    
+    # Current epoch time
+    CurrentStamp = str(time.time()).split('.')[0]
```
Comment (stapelberg) [inline contrib/tubestatusdef.py]: Use `CurrentStamp = int(time.time())` instead.

File: contrib/tubestatusdef.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+ check if temp file exists, if not, write it with time stamp in it, else pull time stamp
+ check time stamp vs. poll interval, if it's less than time + interval, skip it, else run and write the temp file again
+ two functions at least, one for time, one for polling, returning status value
+'''
+
+def Throttle(PollInterval,Throttle,TFileName):
+
+    # Does the temporary file exist
+    TFileStatus = os.path.isfile(TFileName)
```
Comment (stapelberg) [inline contrib/tubestatusdef.py]: Why introduce a variable that you only use once? You can use `if os.path.isfile(TFileName)` below, which is more clear when reading.

File: contrib/tubestatusdef.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        TFile.closed
+ 
+        # Set the Remainder high to force the next run
+        Remainder = 1000000 
+
+    # The poll interval in seconds
+    PollInterval = PollInterval * 60
+
+    # If the remainder is less than the poll interval don't run the command, if it isn't run the command
+    if ( Remainder < PollInterval ):
```
Comment (stapelberg) [inline contrib/tubestatusdef.py]: More idiomatic:

```python
if Remainder < PollInterval:
  return False
# …
return True
```

File: contrib/tubestatusdef.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+def RetrieveTFLData(Line,Run,SFileName):
+
+    # TFL Unified API URL
+    TFLDataURL = "https://api.tfl.gov.uk/Line/" + Line + ("/Status?detail=False"
+    "&app_id=&app_key=")
+
+    if Run: 
+        # Read all the information from JSON at the specified URL
+        RawData = urlopen(TFLDataURL).readall().decode('utf8') or die("Error, failed to "
```
Comment (stapelberg) [inline contrib/tubestatusdef.py]: Are you aware of http://docs.python-requests.org/en/latest/? That module might make the code here more readable/shorter.

File: contrib/tubestatusdef.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        TFLData = json.loads(RawData)
+
+        # Sanitize the data to get the line status
+        Scratch = (TFLData[0]['lineStatuses'])
+        LineStatusData = (Scratch[0]['statusSeverityDescription'])
+        
+        # Cache the staus in a file
+        with open(SFileName, 'w+') as SFile:
+            SFile.write(LineStatusData)
+        SFile.closed
```
Comment (stapelberg) [inline contrib/tubestatusdef.py]: Here and below, you don’t need to call `.closed` — when using `with`, the file is automatically closed as soon as you leave the scope of the `with` block. The python docs merely call `.closed` to illustrate that the file is indeed closed. In your code, it doesn’t have any effect.

File: contrib/tubestatusdef.py (hunk 1 above)
Comment (brioscaibriste) [inline contrib/tubestatusdef.py]: I've tried to change this to using int rather than strings but can't quite get it to work as I would like.

Comment (stapelberg) [inline contrib/tubestatusdef.py]: Can you give me more details about what doesn’t work (i.e. the code and the error message you get)? Without anything else, I can’t really answer this.

**Comment by Airblader** [comment]
@brioscaibriste @stapelberg What's the status here?

**Comment by brioscaibriste** [comment]
Hi all

I sent a few mails to Transport For London with regards to using my API key
in open source software. They still haven't replied. I will continue to
chase.

Thanks

Conor

**Comment by Airblader** [comment]
@ivernus So… what now? This PR has been open for a long, long time now.

**Comment by ivernus** [comment]
@Airblader I'm more than happy for it to be merged, it was paused for a bit as TFL would never confirm what to do if you wanted to have a key for open source software. I put in the documentation that an API key is required so it should be good to merge.

**Comment by Airblader** [comment]
Thanks for the answer. There's still a review comment you haven't addressed. Also before we can merge it you'll need to squash the commits so we have a clean git log. But before you do so @stapelberg should perhaps say what his thoughts on it are now. :)

**Comment by stapelberg** [comment]
Given how big the source of this contrib script is, I wonder whether it’ll be easier if it was maintained in a third-party repository. That way changes could be made more quickly and you’re not tied to the i3status release cycle.

Thoughts? I.e., is there any specific reason why this should be bundled with i3status?

**Comment by ivernus** [comment]
that would work @stapelberg , I made it specifically for i3status really as I use that most of the time, so I thought it would be good bundled in. But I'm happy to just keep it separate if you think it would be better.

**Comment by stapelberg** [comment]
Thanks for the update.

I think it might be best to put it into a repository owned by you, possibly called tubestatus-for-i3status (like there is quickswitch-for-i3, for example).

I’ll close the PR.

**Pull request closed without merging**